flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java

```diff
@@ -0,0 +1,214 @@
+                        // Wait for job to be finished
+                        .thenCompose(
+                                path ->
+                                        executionGraph
+                                                .getTerminationFuture()
+                                                .thenApply(
+                                                        status ->
+                                                                new JobStatusAndSavepointPath(
+                                                                        status, path)))
+                        // check that the job finished successfully
+                        .handleAsync(
+                                this::handleStopWithSavepointOperationResult,
+                                context.getMainThreadExecutor());
+
+        FutureUtils.forward(savepointFuture, operationCompletionFuture);
+    }
+
+    private String handleSavepointOperationResult(
+            CompletedCheckpoint completedCheckpoint, Throwable throwable) {
+        if (throwable != null) {
+            context.runIfState(this, () -> handleAnyFailure(throwable));
+            throw new SavepointCreationException(
+                    "Unable to stop with savepoint. Error while creating the savepoint.",
+                    throwable);
+        }
+        return completedCheckpoint.getExternalPointer();
+    }
+
+    private String handleStopWithSavepointOperationResult(
+            JobStatusAndSavepointPath statusAndPath, Throwable throwable) {
+        // filter out exception from first stage to show correct root cause
+        if (throwable != null) {
+            if (throwable instanceof SavepointCreationException) {
+                throw (SavepointCreationException) throwable;
+            } else {
+                throw new CompletionException(
+                        "Error while stopping job after creating savepoint", throwable);
+            }
+        }
+
+        if (statusAndPath.getStatus() == JobStatus.FINISHED) {
+            // ensure that all vertices from the execution graph are in state FINISHED. There might
+            // be cases where only some subgraphs of the job are finished.
+            if (!areAllVerticesFinished()) {
+                if (getLogger().isDebugEnabled()) {
+                    getLogger().debug("Printing vertex states:");
+                    for (Execution e : getExecutionGraph().getRegisteredExecutions().values()) {
+                        getLogger().debug("{} = {}", e.getVertex(), e.getState());
+                    }
+                }
+                final FlinkException exception =
+                        new FlinkException(
+                                "Not all vertices of job are finished. Stop with savepoint operation failed.");
+                context.runIfState(this, () -> handleAnyFailure(exception));
+                throw new CompletionException("Unable to stop with savepoint", exception);
```

Comment:
here again we have one path for the failure handling and another path for communicating the result back to the user. I think this should be one and the same path. If we leave this state in any other state than `Finished`, then we should fail the operation.